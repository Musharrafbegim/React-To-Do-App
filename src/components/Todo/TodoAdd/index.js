import React, { Fragment } from 'react'
import PropTypes from 'prop-types'
import { Button } from '@blueprintjs/core'
import styles from './index.module.scss'

const TodoAdd = ({ toggleView, hasTasks }) => {
  const message = `${
    hasTasks
      ? 'Nice! Looks like you have some stuff to do, but you can have more!'
      : "Nothing to do yet? Think about it and let's start!"
  }`

  return (
    <Fragment>
      <h1>Hi!</h1>
      <p data-test="message" className={styles.description}>
        {message}
      </p>
      <Button
        large
        minimal
        data-test="button"
        className="panel-button"
        onClick={toggleView}
      >
        Add Task
      </Button>
    </Fragment>
  )
}

TodoAdd.defaultProps = {
  hasTasks: false,
}
TodoAdd.propTypes = {
  toggleView: PropTypes.func.isRequired,
  hasTasks: PropTypes.bool,
}

export default TodoAdd
