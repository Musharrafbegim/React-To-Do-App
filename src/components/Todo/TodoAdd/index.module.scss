@import 'styles/colors.scss';

.description {
  font-weight: 300;
  font-size: 1.2rem;
  color: $outline-color;
  margin-bottom: 20px;
}
