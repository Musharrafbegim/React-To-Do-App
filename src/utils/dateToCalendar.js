import { format, isToday, isTomorrow, isThisWeek, isThisYear } from 'date-fns'

export default date => {
  if (isToday(date)) return 'Today'
  if (isTomorrow(date)) return 'Tomorrow'
  if (isThisWeek(date)) return format(date, 'dddd')
  return format(date, `MMM Do${!isThisYear(date) ? ', YYYY' : ''}`)
}
