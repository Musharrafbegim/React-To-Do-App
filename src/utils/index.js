export { default as dateToCalendar } from './dateToCalendar'
export { default as groupByDate } from './groupByDate'
