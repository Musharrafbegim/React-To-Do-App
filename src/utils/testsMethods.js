import checkPropTypes from 'check-prop-types'
import { createStore } from 'redux'
import reducers from 'store/reducers'

export const findByTestAttr = (component, attr) =>
  component.find(`[data-test='${attr}']`)

export const checkProps = ({ propTypes, name }, expectedProps) =>
  checkPropTypes(propTypes, expectedProps, 'props', name)

export const testStore = initialState => createStore(reducers, initialState)
