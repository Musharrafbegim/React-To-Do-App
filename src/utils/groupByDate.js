import dateToCalendar from './dateToCalendar'

export default (list, filterBy) =>
  list.filter(filterBy).reduce((r, acc) => {
    const key = dateToCalendar(acc.date)
    const obj = r
    obj[key] = [...(obj[key] || []), acc]
    return obj
  }, {})
