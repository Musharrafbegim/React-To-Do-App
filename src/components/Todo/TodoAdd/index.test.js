import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { findByTestAttr, checkProps } from 'utils/testsMethods'
import TodoAdd from '.'

const setUp = props => shallow(<TodoAdd {...props} />)

describe('TodoAdd Component', () => {
  describe('Check PropTypes', () => {
    it('should not throw a warning', () => {
      const expectedProps = {
        toggleView: () => {},
        hasProps: false,
      }

      const propsErr = checkProps(TodoAdd, expectedProps)
      expect(propsErr).toBeUndefined()
    })
  })

  let wrapper
  let mockToggleView

  beforeEach(() => {
    mockToggleView = jest.fn()
    const props = {
      toggleView: mockToggleView,
    }
    wrapper = setUp(props)
  })

  describe('Rendering', () => {
    it('should render without errors', () => {
      expect(toJson(wrapper)).toMatchSnapshot()
    })

    it('should render Add Task button', () => {
      const button = findByTestAttr(wrapper, 'button')

      expect(button.length).toBe(1)
      expect(
        button
          .children()
          .first()
          .text(),
      ).toBe('Add Task')
    })
  })

  describe('Interactions', () => {
    it('should emit toggle view on click event', () => {
      const button = findByTestAttr(wrapper, 'button')
      button.simulate('click')

      expect(mockToggleView).toHaveBeenCalled()
    })

    test('should show text to encourage to add a task if there are no tasks yet', () => {
      const p = findByTestAttr(wrapper, 'message')
      const expectedText = "Nothing to do yet? Think about it and let's start!"

      expect(p.text()).toEqual(expectedText)
    })

    test('should notify user when he has tasks', () => {
      const props = {
        toggleView: mockToggleView,
        hasTasks: true,
      }
      wrapper = setUp(props)
      const p = findByTestAttr(wrapper, 'message')
      const expectedText =
        'Nice! Looks like you have some stuff to do, but you can have more!'

      expect(p.text()).toEqual(expectedText)
    })
  })
})
