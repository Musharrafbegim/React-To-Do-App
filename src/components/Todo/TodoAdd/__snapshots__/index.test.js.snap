// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoAdd Component Rendering should render without errors 1`] = `
<Fragment>
  <h1>
    Hi!
  </h1>
  <p
    className="description"
    data-test="message"
  >
    Nothing to do yet? Think about it and let's start!
  </p>
  <Blueprint3.Button
    className="panel-button"
    data-test="button"
    large={true}
    minimal={true}
    onClick={[MockFunction]}
  >
    Add Task
  </Blueprint3.Button>
</Fragment>
`;
